src/mouse_mode.h

```diff
@@ -404,6 +404,88 @@ class RideTileDataMouseMode : public TileDataMouseMode<TileData> {
 };
 
 typedef TileDataMouseMode<CursorTileData> CursorMouseMode; ///< Mouse mode displaying a cursor of some size at the ground.
-typedef RideTileDataMouseMode<RideTileData> RideMouseMode; ///< Mouse mode displaying a cursor and (part of) a ride.
+
+/** Mouse mode displaying a cursor and (part of) a ride. */
+class RideMouseMode : public VoxelTileDataMouseMode<VoxelTileData<VoxelRideData>> {
+public:
+	RideMouseMode() : VoxelTileDataMouseMode<VoxelTileData<VoxelRideData>>()
+	{
+	}
+
+	~RideMouseMode()
+	{
+	}
+
+	bool GetRide(const Voxel *voxel, const XYZPoint16 &voxel_pos, SmallRideInstance *sri, uint16 *instance_data) override
+	{
+		uint32 index = this->GetTileIndex(voxel_pos.x, voxel_pos.y);
+		if (index == INVALID_TILE_INDEX) return false;
+
+		const VoxelTileData<VoxelRideData> &td = this->tile_data[index];
+		if (!td.cursor_enabled || voxel_pos.z < td.lowest || voxel_pos.z > td.highest) return false;
+
+		const VoxelRideData &vrd = td.ride_info[voxel_pos.z - td.lowest];
+		*sri = vrd.sri;
+		*instance_data = vrd.instance_data;
+		return true;
+	}
+
+	/**
+	 * Set ride data at the given position in the area. Tiles with disabled cursor are silently skipped.
+	 * @param pos World position.
+	 * @param sri Ride instance number.
+	 * @param instance_data Instance data.
+	 */
+	void SetRideData(const XYZPoint16 &pos, SmallRideInstance sri, uint16 instance_data)
+	{
+		VoxelTileData<VoxelRideData> &td = this->GetTileData(pos);
+		if (!td.cursor_enabled) return;
+		VoxelRideData &vrd = td.ride_info[pos.z - td.lowest];
+		vrd.sri = sri;
+		vrd.instance_data = instance_data;
+	}
+};
+
+/** Mouse mode displaying a cursor and fences. */
+class FencesMouseMode : public VoxelTileDataMouseMode<VoxelTileData<VoxelFenceData>> {
```

Comment:
Maybe "EdgeMouseMode" ?i can't think of anything at the moment, but it seems reasonable that there might be something other than fences that's placed on the edge of tiles